Update `userLists`, `web3listener`, and `settings` Redux files to use TypeScript

marco opened the pull request:
Here is some initial work on converting some files to TypeScript. I based most of the changes off of the patterns I saw in #2495.

As far as I can tell, these changes don't break any `yarn lint` rules, but it's definitely possible I'm missing something!

jinchung commented on src/redux/settings.ts:
```diff
@@ -102,15 +155,15 @@ export const settingsChangeNativeCurrency = nativeCurrency => async dispatch =>
   chainId: 1,
   language: 'en',
   nativeCurrency: 'USD',
   network: networkTypes.mainnet,
```
we can use the enum: `Network.mainnet` and then remove the import of `networkTypes`

jinchung commented on src/redux/settings.ts:
```diff
@@ -25,7 +26,51 @@ const SETTINGS_UPDATE_NETWORK_SUCCESS =
+  accountAddress: string;
+  chainId: number;
+  language: string;
+  nativeCurrency: string;
```
this is a great point! let's make a separate linear ticket for the `language` and `nativeCurrency` types so we can keep track of it without it blocking this PR!

jinchung commented on src/redux/userLists.ts:
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
   dispatch({ type: USER_LISTS_LOAD_REQUEST });
   try {
-    const defaultLists = DefaultTokenLists[network] || [];
-    const userLists = await getUserLists(network);
```
yes you're right - good catch!

jinchung commented on src/redux/userLists.ts:
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
+
+export const userListsLoadState = () => async (
+  dispatch: AppDispatch,
+  getState: () => AppState
```
You can import `AppGetState` and replace `() => AppState` with `AppGetState`

jinchung replied:
Same with all the other places using `() => AppState`

jinchung commented on src/redux/userLists.ts:
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
+interface UserListsLoadSuccessAction {
+  type: typeof USER_LISTS_LOAD_SUCCESS;
+  payload: {
+    lists: UserList[];
```
this isn't part of your TS changes, but noticed we can improve this: the payload could be the `UserList[]` directly instead of being an object with `lists` inside of it. This also means updating line 115 and line 241 accordingly

jinchung commented:
on line 125: 
```      dispatch({
        payload: true,
        type: USER_LISTS_READY,
      });```
the `payload: true` is not needed

jinchung commented on src/redux/web3listener.ts:
```diff
@@ -1,11 +1,15 @@
-const updateMulticall = blockNumber => async (dispatch, getState) => {
+const updateMulticall = (blockNumber: number) => async (
+  dispatch: AppDispatch,
+  getState: () => AppState
```
`AppGetState` usage in this file as well

jinchung commented on src/redux/web3listener.ts:
```diff
@@ -22,11 +26,17 @@ const debouncedUpdateMulticallListeners = debounce(blockNumber => {
-export const web3ListenerInit = () => {
-  web3Provider.pollingInterval = 10000;
-  web3Provider.on('block', debouncedUpdateMulticallListeners);
+let typedWeb3Provider = web3Provider as JsonRpcProvider | null;
```
ah ideally `handlers/web3` should be TS'd. I realize now that a lot of the items in `handlers` will be used even in the `redux` files so it may be beneficial to include these as part of your conversion

marco replied:
I can start working on `handlers/web3` tomorrow morning!

marco replied:
@jinchung `handlers/web3` ended up being relatively involved, so I opened #2511 to track that separately. Thanks!

jinchung commented on src/redux/userLists.ts:
```diff
@@ -21,16 +23,90 @@ const USER_LISTS_UPDATE_LISTS = 'userLists/USER_LISTS_UPDATE_LISTS';
+      (DefaultTokenLists as Record<
+        Network,
+        typeof DefaultTokenLists[keyof typeof DefaultTokenLists]
+      >)[network] || [];
```
this is a good point - `Network` is actually a bit of a confusing type - we need to actually split this out into 3 different types used for different purposes; can discuss further later

jinchung replied:
in the interim, can this type be exposed in `references` itself where `DefaultTokenLists` comes from?

marco replied:
Okay, for now I moved this type into `references` and now import it when doing the cast, and in the future we can split up `Network` like you referenced. Thanks!

jinchung commented on src/entities/userLists.ts:
```diff
@@ -0,0 +1,10 @@
+ * A single user list, with fields based on the data found in
+ * `src/references/default-token-lists.json`.
+ */
+export interface UserList {
```
we have a pattern of exporting to `entities/index` (you can check out the other ones that are already in there) and only importing from `entities` instead of having to do `entities/userList` etc!

marco replied:
Great, thanks! I pushed an update that should address this and your second comment. I'll also add the `handlers/web3` updates when I'm done with them later today.

markdalgleish commented on src/redux/settings.ts:
```diff
@@ -25,7 +26,51 @@ const SETTINGS_UPDATE_NETWORK_SUCCESS =
+  payload: SettingsState['language'];
+}
+
+export const settingsLoadState = () => async (dispatch: AppDispatch) => {
```
Same issue I noticed here: https://github.com/rainbow-me/rainbow/pull/2505#discussion_r743664238

tl;dr `AppDispatch` is being resolved as `AppDispatch<any>`, so it might be worth explicitly typing the actions here with `Dispatch<SettingsStateUpdateAction>`.

The tricky part is that some actions being dispatched in this file are coming from JS files so you might want to annotate those dispatches with `// @ts-expect-error` comments — that way you'll be prompted to come back and clean up this file when those files are converted to TS.

marco commented:
Rebased and updated this one now that the the `web3` changes have been merged.

Proof of work:

<img width="200" alt="Screen Shot 2021-12-07 at 11 20 39 AM" src="https://user-images.githubusercontent.com/17259768/145082970-18eecf43-e3cc-4d8c-8da6-ea2800d57d9f.png">

jinchung commented on src/redux/web3listener.ts:
```diff
@@ -1,11 +1,15 @@
 // -- Actions ---------------------------------------- //
-const updateMulticall = blockNumber => async (dispatch, getState) => {
+const updateMulticall = (blockNumber: number) => async (
+  dispatch: ThunkDispatch<AppState, unknown, never>,
```
why ThunkDispatch instead of AppDispatch ?

jinchung replied:
same for the other redux files

marco replied:
This relates to https://github.com/rainbow-me/rainbow/pull/2498#discussion_r744327085, since there are other JS reducers that get included, the standard `AppDispatch` actually doesn't end up doing any type checking. Basically it resolves to `Dispatch<any>`. In this case I think `ThunkDispatch` instead of `Dispatch` because `multicallUpdateOutdatedListeners(blockNumber)` is being passed in instead of an object. For most of the other ones though, I used a `Dispatch` and specified which the valid actions are in order to get the type checking.

jinchung commented on src/redux/userLists.ts:
```diff
@@ -82,9 +160,11 @@ export const userListsClearList = listId => (dispatch, getState) => {
+    allNewLists[listIndex] = newList;
+  }
 
   dispatch({
```
we can move this dispatch and line 173 (saveUserLists) into the if statement above

marco replied:
Okay, I updated this dispatch as well as the one below to be in the `if` statements, thanks!

jinchung commented on src/redux/userLists.ts:
```diff
@@ -114,19 +198,21 @@ export const userListsUpdateList = (assetAddress, listId, add = true) => (
+      allNewLists[listIndex] = newList;
+    }
 
     dispatch({
```
same thing we can move dispatch and saveUserLists into the if as well

markdalgleish commented on src/redux/userLists.ts:
```diff
@@ -114,44 +198,46 @@ export const userListsUpdateList = (assetAddress, listId, add = true) => (
-    if (add) {
-      dispatch(emitAssetRequest(assetAddress));
-    }
+    if (listIndex) {
```
`listIndex` could be `0`, right? Should this be `if (listIndex !== null)`?

jinchung replied:
ah yes very good point - nice catch!

marco replied:
Yes definitely, I updated that as well as the one below, thanks!

markdalgleish commented on src/redux/userLists.ts:
```diff
@@ -82,20 +160,26 @@ export const userListsClearList = listId => (dispatch, getState) => {
-  const newList = { ...allNewLists[listIndex] };
-  newList.tokens = [];
-  allNewLists[listIndex] = newList;
+  if (listIndex) {
```
Same here: `if (listIndex !== null)`?

jinchung reviewed (approved):
🥪

markdalgleish reviewed (approved):
🎉
